### PR TITLE
chore(C): add circular linked list

### PR DESCRIPTION
**Pull request by risanlang**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] New algorithm
- [ ] Optimization in previous algorithms
- [ ] Code style update
- [ ] Refactor
- [ ] Documentation
- [ ] Other, please describe:

**Briefly describe the changes in this PR**

**Other information:**
Added Circular Linked List Program in C language

### CODE REVIEW

**Comment by risanlang** [comment]
@ming-tsai sir i am still new to github and i am trying to learn to contribute to Open Source project however i don't understand what is this error means
I would really appreciate if someone would help me understand this error

Thanks..
Screenshot from 2021-05-28 08-52-31

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c [hunk 1]
```diff
@@ -0,0 +1,92 @@
+
+//Circular Linked List
+//insertion and Deletion at the begining of the Linked List
```
Comment (Ashborn-SM) [inline algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:3]: you have spelling mistake here `begining`

**Comment by Ashborn-SM** [comment]
> **What kind of change does this PR introduce?** (check at least one)
> 
> * [ ]  Bugfix
> * [ ]  New algorithm
> * [ ]  Optimization in previous algorithms
> * [ ]  Code style update
> * [ ]  Refactor
> * [ ]  Documentation
> * [ ]  Other, please describe:
> 
> **Briefly describe the changes in this PR**
> 
> **Other information:**

You have to check atleast one of these option and write something brief about what it is about.

For example:
This pr introduces a new algorithm put an x inside the brackets with no spaces like ` [x]`

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c (hunk 1 above)
Comment (risanlang) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c]: Beginning spelling corrected
Screenshot from 2021-05-28 14-52-46

**Comment by risanlang** [comment]
@Ashborn-SM  sir do i need to make a new pull request to add an [x] New algorithm ?

**Comment by Ashborn-SM** [comment]
> @Ashborn-SM sir do i need to make a new pull request to add an [x] New algorithm ?

no need for that, all you have to do is edit the comment.
Also try to give some brief description about the the pr, it helps to understand what this is about quite easily

**Comment by risanlang** [comment]
> > @Ashborn-SM sir do i need to make a new pull request to add an [x] New algorithm ?
> 
> no need for that, all you have to do is edit the comment.
> Also try to give some brief description about the the pr, it helps to understand what this is about quite easily

sir i've added a brief description about the pr just like how you told me to
Thanks

**Comment by Ashborn-SM** [comment]
> > > @Ashborn-SM sir do i need to make a new pull request to add an [x] New algorithm ?
> > 
> > 
> > no need for that, all you have to do is edit the comment.
> > Also try to give some brief description about the the pr, it helps to understand what this is about quite easily
> 
> sir i've added a brief description about the pr just like how you told me to
> Thanks

Description must be about the pr not about the algorithm.
Here are some closed pr's refer this - https://github.com/MakeContributions/DSA/pulls?q=is%3Apr+is%3Aclosed

**Comment by risanlang** [comment]
@Ashborn-SM  sorry i thought i have to add a description about the algorithm but instead its for the PR 
I did add the description as instructed

**Comment by Ashborn-SM** [comment]
@risanlang Good work.
Next time give description under `Briefly describe the changes in this PR: `. `Other infromation:` is information related to your pr or your code.

Now update the readme as well.

**Comment by Ashborn-SM** [comment]
@risanlang Good work.
Next time give description under `Briefly describe the changes in this PR:` , `Other infromation:` is some additional information related to your pr or your code, if you have none leave it blank

Now update the readme as well.

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ 
+}
+
+//Deletion at the Beginning of the Circular Linked List
+
+void deleteBeg(){
+  //NO Deletion when list is empty
+   if(head == NULL) {
+    printf("List is Empty , deletion not possible") ;
+   }
```
Comment (Arsenic-ATG) [inline algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:67]: you would not want to continue the execution of the function in case the list is empty. 

```suggestion
   if(head == NULL) {
    printf("List is Empty , deletion not possible") ;
    return;
   }
```

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  //Store head location in a temporary Node pointer.
+  struct node *temp = head ;
```
Comment (Arsenic-ATG) [inline algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:69]: What will happen when there is only one element left in the list ?

**Review by Arsenic-ATG — CHANGES_REQUESTED** [review]
fix the `deleteBeg()` function ( see my comments on source for the same ) 

Also don't forget to add your creation to the [index readme](https://github.com/MakeContributions/DSA/blob/main/algorithms/C/README.md) ( as already pointed out by @Ashborn-SM )

If confused or stuck then feel free to ask questions here.

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c (hunk 2 above)
Comment (risanlang) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:67]: sir how can i return when the return type of the delete function is void
maybe i should use the exit statement is this fine??

Comment (Ashborn-SM) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:67]: True void function cannot return any value but we can use return statement here to terminate the functions execution.

Comment (risanlang) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:67]: Oh ok sir understood i will do that suggested change
Thank you!

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c (hunk 3 above)
Comment (risanlang) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:69]: if there is only one element left in the list then both the head and the tail pointer will point to the that same Node
then we create a temporary pointer [temp] and and make it point to the head
we delete the node pointed by temp
next we make the head point to head->next which is null
Now, head will point to NULL and temp will be deleted
I don't see what is the problem here sir??

Comment (Arsenic-ATG) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:69]: >we delete the node pointed by temp
next we make the head point to head->next which is null

here is the problem,
`free()` just deallocates the memory and doesn't make the pointer point to `NULL`, the pointer will still be pointing to that particular memory location only, it's just that that now that memory location would not be accessible to your program anymore and an attempt to do so ( which will happen when you try to print the contents via `printList()` function ) will lead to [undefined behaviour](https://en.wikipedia.org/wiki/Undefined_behavior) ( you can see this happening when you attempt to perform more deletions than the size of the list ) , such pointers are known as [dangling pointers](https://en.wikipedia.org/wiki/Dangling_pointer) and best way to avoid it is to handle the case when there is only one element in the list separately and make the head manually point to `NULL` after deleting it.

you can do something like this 
```c
if(head == tail)
  {
      free(head);
      head = NULL;
      return;
  }
```

---

Also you might have seen till now that your `printList()` function is also unsafe as it will not check if the list is empty or not and will still try to access the list even if it's empty ( head is pointing to `NULL` ), you might wanna change that.

soemthing like this 
```c
if(head == NULL)
  {
      printf("\nList is empty");
      return;
  }
```

Comment (Ashborn-SM) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:69]: You can actually test it for yourself, try inserting only one element to the list and call `deleteBeg()` and then `printList()`
you will a get a `segmentation fault` error or an unexpected behaviour

Comment (risanlang) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c]: @Ashborn-SM @Arsenic-ATG  Thank you sir for pointing my mistake, I feel like solving a problem on Hackerrank where i pass the test case but when i submit the code i fall short because i did not handle the corner case haha
I have also updated the README.md file as you suggested
if there is any other mistake please do tell me
Thank you again for guiding me

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+   if(head == NULL) {
+    printf("List is Empty , deletion not possible") ;
+    return;
+   }
+   //when both the head and the tail pointer points to the same location
+   //i.e there is only one node left in the list
+   else if(head == tail)
+  {
+      free(head);
+      free(tail);
```
Comment (Ashborn-SM) [inline algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:85]: Here this will throw a `double free error` , it occurs when you use `free()` more than once on the same memory address.
So here both `head` and `tail` both of them point to the same memory address, when you use `free(head)` everything in that memory address is released and since `tail` is pointing to this **released** memory address if you use `free(tail)` it will crash and throw in the error.

suggested changes
```C
else if(head == tail){
    free(head);
    head = NULL;
    tail = NULL;
}
```

Comment (Ashborn-SM) [inline algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c:85]: Here this will throw a `double free error` , it occurs when you use `free()` more than once on the same memory address.
So here both `head` and `tail` both of them point to the same memory address, when you use `free(head)` everything in that memory address is released and since `tail` is now pointing to this **released** memory address if you use `free(tail)` it will crash and throw in the error.

suggested changes
```C
else if(head == tail){
    free(head);
    head = NULL;
    tail = NULL;
}
```

**Review by Ashborn-SM — CHANGES_REQUESTED** [review]

File: algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c (hunk 4 above)
Comment (risanlang) [reply algorithms/C/linked-lists/Insert-and-del-beginning-circular-ll.c]: right 
sorry i will make the changes immediately

**Review by Ashborn-SM — APPROVED** [review]
Good work @risanlang

**Review by Arsenic-ATG — APPROVED** [review]

**Comment by risanlang** [comment]
@Ashborn-SM  sir what do I need to do next?? How do i proceed from here
I would be grateful if you tell me the next step
Thanks!!
Screenshot from 2021-06-01 11-42-06

**Comment by Arsenic-ATG** [comment]
> @Ashborn-SM sir what do I need to do next?? How do i proceed from here
> I would be grateful if you tell me the next step
> Thanks!!

Nothing, your PR has been approved and soon be merged in main by @ming-tsai 
well done 🏅👍

**Comment by risanlang** [comment]
Thank You!! @Arsenic-ATG @Ashborn-SM for helping me in taking the first step in Open source contribution
Now i also realize why git is widely used
